[GTK] Fix build on Clang

ItsJamie9494 opened the pull request:
#### b523d24763260206383aa96158d3bd63685e0dbf
<pre>
[GTK] Fix build on Clang
<a href="https://bugs.webkit.org/show_bug.cgi?id=300332">https://bugs.webkit.org/show_bug.cgi?id=300332</a>

Reviewed by Adrian Perez de Castro.

Update how the baseURL creation and path checking works to avoid g_str_has_* functions,
preventing any unsafe buffer usage

* Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
(WebKit::WebExtension::WebExtension):

Canonical link: <a href="https://commits.webkit.org/301160@main">https://commits.webkit.org/301160@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58e492097b688cd63f26fe33d90acbf505078f0d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44782 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35521 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131963 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45473 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53346 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/95243 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128069 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36299 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/111885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75785 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/35204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75443 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106064 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30277 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134642 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51926 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39721 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103710 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108105 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103481 "Found 1 new API test failure: WebKitGTK/TestWebKitWebExtension:/webkit/WebKitWebExtension/creation-from-directory (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26346 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48829 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27119 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48979 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51818 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57602 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51185 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54541 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52875 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -45,6 +45,7 @@ WebExtension::WebExtension(GFile* resourcesFile, RefPtr<API::Error>& outError)
+    WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
     GUniquePtr<char> baseURL(g_file_get_path(resourcesFile));
     if (!g_str_has_prefix(baseURL.get(), "file"))
         baseURL.reset(g_strconcat("file:", baseURL.get(), nullptr));
```
What you want to do here is: `g_file_get_uri()`. That will return a file:// URL. Then you don't need `g_str_has_prefix()`.

mcatanzaro commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
```diff
@@ -54,6 +55,7 @@ WebExtension::WebExtension(GFile* resourcesFile, RefPtr<API::Error>& outError)
     if (!g_str_has_suffix(baseURL.get(), "/"))
         baseURL.reset(g_strconcat(baseURL.get(), "/", nullptr));
     m_resourceBaseURL = URL { String::fromUTF8(baseURL.get()) };
```
Also: since you used `g_file_get_path()` instead of `g_file_get_uri()`, `baseURL` is in filesystem encoding, which may not be UTF-8. Create it using `stringFromFilesystemRepresentation` instead of `String::fromUTF8`.

(Except don't, because you'll switch to `g_file_get_uri()`, dodging this problem. :)

mcatanzaro commented on Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
I think you can replace the `RELEASE_ASSERT()` with:

```
if (!m_resourceBaseURL.hasPath())
    m_resourceBaseURL.setPath("/"_s);
```

Then you don't need the `g_str_has_suffix()`.

mcatanzaro reviewed (changes requested):
Adding `WTF_ALLOW_UNSAFE_BUFFER_USAGE` is OK when we don't have time to fix the code, but in this case it should be easy enough to rewrite the code to avoid these warnings.

Normally the solution is to convert to `WTF::String` and use that rather than continue to operate directly on the C string. But in this case, even that isn't needed.

ItsJamie9494 committed:
[GTK] Fix build on Clang
https://bugs.webkit.org/show_bug.cgi?id=300332

Reviewed by Adrian Perez de Castro.

Update how the baseURL creation and path checking works to avoid g_str_has_* functions,
preventing any unsafe buffer usage

* Source/WebKit/UIProcess/Extensions/gtk/WebExtensionGtk.cpp:
(WebKit::WebExtension::WebExtension):

Canonical link: https://commits.webkit.org/301160@main
